### PR TITLE
Bug 1797600 - Create cookie banner cfr

### PR DESCRIPTION
**Pull request by iorgamgabriel**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1797600

### CODE REVIEW

**Comment by Amejia481** [comment]
LGTM 👍🏽

**Comment by iorgamgabriel** [comment]
# Request for data collection review form

All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.

1. What questions will you answer with this data?
  * How users interact with the cookie banners handling feature.
2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements? 
  * This data will help us understand how the cookie banners handling feature is used.
3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
  * There are no other alternatives. 
4. Can current instrumentation answer these questions?
  * No. 
5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki. 
Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.

Measurement Description | Data Collection Category | Tracking Bug #
-- | -- | --
Cookie banner cfr was shown to the user  | Category 2 - interaction data |https://bugzilla.mozilla.org/show_bug.cgi?id=1797600

6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
  * https://dictionary.telemetry.mozilla.org/apps/focus_android
7. How long will this data be collected? Choose one of the following:
  * Until version 119. Renewal then may be decided.
8. What populations will you measure?
  * All channels, all locales, all countries  that the feature is active for.
9. If this data collection is default on, what is the opt-out mechanism for users?
  * Default Glean SDK opt-out mechanism. 
10. Please provide a general description of how you will analyze this data.
  * Glean and Amplitude. 
11. Where do you intend to share the results of your analysis?
  *  Only on Glean, Amplitude, and with mobile teams. 
12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
   * No third-party tools.

**Comment by rocketsroger** [comment]
> # Request for data collection review form
> 
> All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.
> 
>     1. What questions will you answer with this data?
> 
> 
>     * How users interact with the cookie banners handling feature.
> 
> 
>     2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements?
> 
> 
>     * This data will help us understand how the cookie banners handling feature is used.
> 
> 
>     3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
> 
> 
>     * There are no other alternatives.
> 
> 
>     4. Can current instrumentation answer these questions?
> 
> 
>     * No.
> 
> 
>     5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.    
>        Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.
> 
> 
> Measurement Description 	Data Collection Category 	Tracking Bug #
> Cookie banner cfr was shown to the user 	Category 2 - interaction data 	https://bugzilla.mozilla.org/show_bug.cgi?id=1797600
> 
>     6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
> 
> 
>     * https://dictionary.telemetry.mozilla.org/apps/focus_android
> 
> 
>     7. How long will this data be collected? Choose one of the following:
> 
> 
>     * Until version 119. Renewal then may be decided.
> 
> 
>     8. What populations will you measure?
> 
> 
>     * All channels, all locales, all countries    that the feature is active for.
> 
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
>     * Default Glean SDK opt-out mechanism.
> 
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
>     * Glean and Amplitude.
> 
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
>     * Only on Glean, Amplitude, and with mobile teams.
> 
> 
>     12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
> 
> 
>     * No third-party tools.

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/apps/fenix )

2) Is there a control mechanism that allows the user to turn the data collection on and off? 

Yes, through the "Send Usage Data" preference in the application settings

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

N/A, collection set to end or be renewed by version 119

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 2, Interaction data

5) Is the data collection request for default-on or default-off?

default-on

6) Does the instrumentation include the addition of **any *new* identifiers**?

No

7) Is the data collection covered by the existing Firefox privacy notice? 

Yes

8) Does the data collection use a third-party collection tool? 

No

## Result

data-review+

File: focus-android/app/src/main/java/org/mozilla/focus/browser/integration/BrowserToolbarIntegration.kt [hunk 1]
```diff
@@ -228,6 +239,7 @@ class BrowserToolbarIntegration(
             setBrowserActionButtons()
             observeEraseCfr()
         }
+        observeCookieBannerCfr()
```
Comment (Amejia481) [inline focus-android/app/src/main/java/org/mozilla/focus/browser/integration/BrowserToolbarIntegration.kt:242]: Should we only observe when the setting is off?

File: focus-android/app/src/main/java/org/mozilla/focus/browser/integration/BrowserToolbarIntegration.kt [hunk 2]
```diff
@@ -283,6 +295,67 @@ class BrowserToolbarIntegration(
...
+                                    onClick = {
+                                        fragment.requireComponents.appStore.dispatch(
+                                            AppAction.OpenSettings(Screen.Settings.Page.CookieBanner),
+                                        )
+                                        onDismissCookieBannerCfr()
+                                    },
+                                )
+                            },
+                        ).apply {
+                            show()
```
Comment (Amejia481) [inline focus-android/app/src/main/java/org/mozilla/focus/browser/integration/BrowserToolbarIntegration.kt:352]: Should we call `stopObserverCookieBannerCfrChanges` after we show the CFR? As it should show once ,right?  Or should it show every time the setting is OFF?

File: focus-android/app/src/main/java/org/mozilla/focus/cfr/CfrMiddleware.kt [hunk 3]
```diff
@@ -40,13 +45,27 @@ class CfrMiddleware(private val appContext: Context) : Middleware<BrowserState,
             }
 
             next(action)
-
+            showCookieBannerCfr(context)
             showTrackingProtectionCfr(action, context)
```
Comment (Amejia481) [inline focus-android/app/src/main/java/org/mozilla/focus/cfr/CfrMiddleware.kt:49]: Here we are checking the `ContentAction.UpdateSecurityInfoAction` with is not relate to cookie banners, if you would like to observe an action that is related to cookie banner check the `CookieBannerAction.UpdateStatusAction`

File: focus-android/app/src/main/res/xml/cookie_banner_settings.xml [hunk 4]
```diff
@@ -1,21 +1,13 @@
...
-<androidx.preference.PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android"
-    xmlns:app="http://schemas.android.com/apk/res-auto">
+<androidx.preference.PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android">
 
     <org.mozilla.focus.cookiebanner.CookieBannerRejectAllPreference
         android:defaultValue="false"
         android:key="@string/pref_key_cookie_banner_reject_all"
         android:layout="@layout/focus_preference_no_icon"
         android:summary="@string/cookie_banner_reject_all_option_summary"
         android:title="@string/cookie_banner_reject_all_option_title" />
```
Comment (Amejia481) [inline focus-android/app/src/main/res/xml/cookie_banner_settings.xml:11]: 👍🏽

File: focus-android/app/src/main/java/org/mozilla/focus/cookiebannerexception/CookieBannerExceptionMiddleware.kt [hunk 5]
```diff
@@ -81,4 +86,18 @@ class CookieBannerExceptionMiddleware(
...
+        withContext(Dispatchers.Main) {
+            appContext.components.engine.clearData(
+                host = domain,
+                data = Engine.BrowsingData.select(
+                    Engine.BrowsingData.AUTH_SESSIONS,
+                    Engine.BrowsingData.ALL_SITE_DATA,
+                ),
+            )
+        }
+    }
```
Comment (Amejia481) [inline focus-android/app/src/main/java/org/mozilla/focus/cookiebannerexception/CookieBannerExceptionMiddleware.kt:102]: 👍🏽

File: focus-android/app/src/main/java/org/mozilla/focus/browser/integration/BrowserToolbarIntegration.kt (hunk 1 above)
Comment (iorgamgabriel) [reply focus-android/app/src/main/java/org/mozilla/focus/browser/integration/BrowserToolbarIntegration.kt]: Done

File: focus-android/app/src/main/java/org/mozilla/focus/browser/integration/BrowserToolbarIntegration.kt (hunk 2 above)
Comment (iorgamgabriel) [reply focus-android/app/src/main/java/org/mozilla/focus/browser/integration/BrowserToolbarIntegration.kt:361]: Done

File: focus-android/app/src/main/java/org/mozilla/focus/cfr/CfrMiddleware.kt [hunk 6]
```diff
@@ -83,4 +102,22 @@ class CfrMiddleware(private val appContext: Context) : Middleware<BrowserState,
...
+
+    private fun shouldShowCookieBannerCfr(context: MiddlewareContext<BrowserState, BrowserAction>): Boolean {
+        return (
+            appContext.settings.shouldShowCookieBannerCfr &&
+                appContext.settings.isCookieBannerEnable &&
+                appContext.settings.getCurrentCookieBannerOptionFromSharePref() ==
+                CookieBannerOption.CookieBannerDisabled() &&
+                context.state.findTabOrCustomTabOrSelectedTab(context.state.selectedTabId)?.cookieBanner ==
+                EngineSession.CookieBannerHandlingStatus.DETECTED
+            )
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply focus-android/app/src/main/java/org/mozilla/focus/cfr/CfrMiddleware.kt]: You find and example here https://github.com/mozilla-mobile/fenix/pull/28405/files#diff-84facb1693d1b185822e99641197460c68ea9bce91c6a438c50111abff9e7da8R494-R510

**Review by Amejia481 — APPROVED** [review]

**Pull request merged**